### ISSUE TITLE
Avoid overwriting

### ISSUE THREAD

**Issue by msdobrescu** [flacon/flacon#190]
Starting from source .cue files that eventually define the same album target paths, even though they are different versions and come from different file sets, the split process does one of the two things:
- either crashes while accessing the same target file for saving from different source files
- either successfully overwrites existing files
Such scenario must be avoided by checking the target path, and if that exists, to create a second path, for instance, by adding a suffix like "(n)", where n is a number, similarly to what various file browsers do.

**Comment by SokoloffA** [comment]
There are 2 situations here:

1. You have already split such an album some time ago. And now you're processing it again. In this case, overwriting files is not a bug, but a feature. For example, if you forgot to correct the tag, you can fix it by simply running the program again.

2. Another situation is that you open 2 CUE files that will overwrite each other's files.
  a. The user does not use tags in the output file template
  b. The user added duplicate album CUE files.

You are right, the program should control items 2.a and 2.b. 

Point 2.a is strange. And most likely, the user needs to configure the template. 

Point 2.b is possible in real life. And in my opinion, the program should output a message like:

> Conversion is not possible:
> Disc 3 "Queen - Miracle" will overwrite the files of disc 1 "Queen - Miracle".

I'm not thrilled with the message, maybe you can suggest a better one

**Comment by msdobrescu** [comment]
Indeed 2. I have some that are actually from different issues, with different number of tracks, but the date + album name are the same, so the problem. Should not proceed to split anyway, unless there is a way to suffix them somehow. Even a GUID suffix is acceptable to me. anything to avoid overriding. It's also possible the target to exist already (as a path).

**Comment by SokoloffA** [comment]
I will add such a check this weekend. And I'll post a beta version with a few changes.
Are you using the Linux or macOS version?

**Comment by SokoloffA** [comment]
I'll let you know when the beta version is ready

**Comment by msdobrescu** [comment]
I use a Gentoo-based system, MocaccinoOS.

**Comment by SokoloffA** [comment]
I found a number of issues with the new functionality from the beta version. Therefore, I have published bug fixes in a stable codebase. Your problem is fixed in [flacon 9.5.1](https://github.com/flacon/flacon/releases/tag/v9.5.1) I planned to check filenames in real time, but there were problems with performance. So now I have added the check when clicking the "Convert" button.

I'm still hoping to add real-time checks in the next major release. If you're still interested, I'll let you know when the beta version is ready. 
I leave the issue open

**Comment by SokoloffA** [comment]
Done in the 10.0.0

**Comment by msdobrescu** [comment]
I'm using Gentoo, I've managed to build v10.0 today, I'll let you know if there are issues with this.

**Comment by SokoloffA** [comment]
@msdobrescu 
Hi! 

> The target paths are the same and the resulting files differ. 

Unfortunately, I don't quite understand what this means, and how it can happen.

Each track has a resultFilePath method, this method returns the full path to the resulting audio file (taking into account the track tags, the "Start num", the output "Direcory", and the "Pattern"). I check that the resultFilePath for all tracks returns unique names.

Please give me an example of your situation.
1. Value of "Direcory" form main window
2. Value of "Pattern" form main window
3. CUE files.

**Comment by SokoloffA** [comment]
> In conclusion, must avoid writing from several sources in the same directory.

The situation is more complicated here. There may be another option:
*  Pink Floyd - The Wall [Disk1]
*  Pink Floyd - The Wall [Disk2]

In this case, we need to allow to write files to one directory. But the first track from the second disc should be numbered as 14 (13 songs of the first disc + 1)

I think I'll have to check,
1. that 2 files with the same number (track number + starting number) are not created in the same directory
2. that no CUE files are overwritten by each other.

**Comment by msdobrescu** [comment]
I would not overcomplicate things, although it's a nice feature.
For your case above I would create two directories, Disk1 & Disk2, each with it's files.
On top of that, if you prefer, I'd do the same as you do now for files also for the last directory in the targeted path.
To be a little evil, imagine you have two sources of the Pink Floyd album above, both with their disc 1 & 2.

**Comment by SokoloffA** [comment]
I personally prefer to have the entire dual disk in one directory. And my desire is above all :)

**Comment by SokoloffA** [comment]
I personally prefer to have the entire dual disk in one directory. And my desire is above all :)

> To be a little evil, imagine you have two sources of the Pink Floyd album above, both with their disc 1 & 2.

In this case, everything should work:
1. At the beginning we have 4 discs, they all look at the same directory and in 4 tracks has the number 1. As a result, we have 4 locks. 
 
2. Then the user changes the starting number for disks No. 2. But the lock will remain, because tracks from different versions have the same numbers. 

3. Now the user can change the album name for one version. There are no locks - now 2 different directories will be created with different versions of the album.

I think it should work.

**Comment by SokoloffA** [comment]
I thought maybe if the first track doesn't start with 1, CUE file should have the name like "Pink Floyd - The Wall - tracks 13-25.cue"

**Comment by msdobrescu** [comment]
Let's have options, then.
- One way to serialize whatever it comes, as you suggest.
- One way to give the same error/warning as today, so the user can handle them manually (like appending something to some tag in order to generate the resulting directories with different naming and decide later what to do with them).

Leaving it as a warning with the possibility to see the list of conflicts would be enough (second option above), then, if the user does nothing about it, the files can be serialized as you propose (first option above).

Note, serializing means to have a merged cue file.

What do you think?

**Comment by SokoloffA** [comment]
Hi! I'm sorry for the long silence, I've been terribly busy these days.

I don't plan to block all the mistakes that a user can make.  Only obvious erroneous situations.
For example, if the files overwrite each other during the conversion, this is a obvious error, we check it. 

Closer to our conversation, a mix of tracks of several versions of the album in one directory, it's unlikely that user wants to get. I can easy check this and show a warning or error message. I'm leaning towards a mistake.

> Note, serializing means to have a merged cue file.

That's not so. People create CUE files to be able to glue original audio files from splitted ones.
If user had several disk at the input, then there should be several after splitting + merging, and for this we need to have several CUE files.

**Comment by msdobrescu** [comment]
I am glad you get closer.
I'd be happy to have at least a warning when files from different sources would land in the same directory path.

**Comment by SokoloffA** [comment]
I added a text like 

![Dialog](https://user-images.githubusercontent.com/854935/226187609-7860c279-334d-4f82-8050-59ead77140af.png)

If there are better suggestions on the text, feel free to offer.

**Comment by msdobrescu** [comment]
Hi, I am not sure what this addresses.
I see two cases:
- when files names are overlapping in the same directory
- when two albums are written in the same directory
The first one was already treated.
How about the second one?

**Comment by SokoloffA** [comment]
I can imagine 3 situations where files from different CUEs end up in the same directory:

1. The disks have completely identical tracks. Example, one album in CD and HD quality.
2. Different editions of the same album, track lists are a bit different. For example, tracks are added, the order of tracks is changed, or some tracks are replaced by others. As an example, the [early Uriah Heep albums](https://en.wikipedia.org/wiki/...Very_%27Eavy_...Very_%27Umble) had a different set of songs in different countries
3. Double albums.

* In situation 1, the program will show an error "Disk XXX will overwrite the files of this disk".
* In situations 2 and 3, the tracks will have the same numbers and the added check will be triggered.
  * In situation 3, the user's choice is either to change the starting track number, or add the "Disk 2" suffix.
  * In situation 2,  the user will either leave one album edition (delete others) or add a suffixes.

I don't see any other situations.

**Comment by msdobrescu** [comment]
Well, 2 & 3 for me are cases to be warned just as in case 1, with regard to the source of the problem, then the user takes action or not - with the risk of mixing things. The encoding takes place in parallel for the tracks, so it is not sure which would end up in that directory. So, for me are three cases, with regard to the resulting files:

1. Same name for the files that are split in an album only that  (badly written CUE or untitled tracks + a particular pattern)
Example: album with all tracks called "Untitled", split pattern %A/%t. (I've met the case)

2. Files from different album sources that have the same final path through its pattern.
Example:
Album "A", tracks "T1", "T2", "T3", stored on a path like "A".
Album "A, with bonus edition", with tracks "T1", "T2", "T3", "T4" stored on a path like "A (+Bonus)"
Using split pattern "%A/%n - %t"

3. User selects two cue files present for the same album in its directory (like indexed and simple versions).
In such case file names and numbers may even differ (but must ensure they end up in a different directory).

Regardless the case, I'd check and warn:
- if there are two files having the same name on the same path;
- if there are two different album sources that generate files on the same path (source cues are on different paths and generate files on the same path);
- if there are two cues from the same album path that would end-up split in the same path.

**Comment by SokoloffA** [comment]
I apologize for the long silence. 
Today I found some time to improve disk validation.

Now we have the following checks:
1. Error if tracks overwrite files (taking into account the template).
2. Error if there are tracks with the same track number and creating files in the same directory. The pattern doesn't matter.
3. Warning if any disks use the same CUE file.
4. Warning if any disks use the same audio file.

[I also added tests for various situations](https://github.com/flacon/flacon/tree/develop/tests/data/testValidator). I think I've covered all your cases.
